numpy/core/src/multiarray/ctors.c

```diff
@@ -2105,11 +2105,31 @@ PyArray_FromStructInterface(PyObject *input)
         }
     }
 
+    /* a tuple to hold references */
+    PyObject *refs = PyTuple_New(2);
+    if (!refs) {
+        Py_DECREF(attr);
+        return NULL;
+    }
+
+    /* add a reference to the object sharing the data */
+    Py_INCREF(input);
+    PyTuple_SET_ITEM(refs, 0, input);
+
+    /* take a reference to the PyCapsule containing the PyArrayInterface
+     * structure. When the PyCapsule reference is released the PyCapsule
+     * destructor will free any resources that need to persist while numpy has
+     * access to the data. */
+    PyTuple_SET_ITEM(refs, 1,  attr);
+
+    /* create the numpy array, this call adds a reference to refs */
     PyObject *ret = PyArray_NewFromDescrAndBase(
             &PyArray_Type, thetype,
             inter->nd, inter->shape, inter->strides, inter->data,
-            inter->flags, NULL, input);
-    Py_DECREF(attr);
+            inter->flags, NULL, refs);
```

Comment:
(Not much of an opinion on whether this should be backported).  But what do you think about my proposal to change this to:
```C
ctx = PyCapsule_GetContext(attr)
if (ctx == NULL && PyErr_Occurred()) {
    /* error cleanup */
    return NULL;
}
/* ctx should not be NULL, but do not rely on it for now: */ 
ref = ctx != NULL ? ctx : input;
```
I feel like that may be a nice way to achieve the same thing, and makes me more comfortable with the change.

The `copy=NEVER` champions could look into adding something a `ctx == input` check if desired.

Reply:
> what do you think about my proposal to change this to 

Two things come to mind. 

Note that the type of the `PyCapsule`'s context field is `void *`, and that the `base` argument to `PyArray_NewFromDescrAndBase`  is type `PyObject*`.
https://github.com/numpy/numpy/blob/1a0d1fba57c76501bd47828ddcafdb924f89eb15/numpy/core/src/multiarray/ctors.c#L948-L951

The proposed change would make memory management of resources other than `PyObject`'s more complicated because it would require the capsule's context field be a `PyObject` but in data shared with numpy from C/C++ the most common use case would be management of temporary buffers allocated with `new` or `malloc`.  See the self contained example I've been using for testing linked to above for an example. With the proposed change, the capsule's context field could no longer be used to track such resources directly, and the capsule's destructor could no longer be used to directly release them. Such resources would instead need to be managed by a secondary object. This could be accomplished, by setting the context field to a second `PyCapsule`, but it is more complicated to do it that way.

There's also a potential type safety issue introduced here because the`PyCapsule`'s context is type `void * `.  It would be easy to misuse this design by inadvertently passing something other than a `PyObject` as the context.  Such accidental misuse would lead to an invalid memory access, and would potentially crash with a segv. There's not a safe way in C to check if the memory pointed to by the `void*` context is in fact a `PyObject*` before numpy uses it. One could argue that it's the responsibility of the sharer of the data, but it's worth noting that this issue isn't present in the original proposal.

I don't know that these two things are necessarily show stoppers for your suggestion, but they might be worth considering.

Reply:
> Note that the type of the PyCapsule's context field is void *, and that the base argument to PyArray_NewFromDescrAndBase is type PyObject*.

Hmm, missed that.  I guess the docs specify it must be an object, but this was never really enforced, so it is unclear that we should do it.
I can also see the point that it would limit your flexibility compared to this solution, have to think a bit more about this, but if others want to move forward, please do not worry too much... I suppose in the end we have "flags" we can use to tag on new information to add new information (not ideal but OK).

A small problem would be if there are consumers other than NumPy, but that would only affect your type actually using this.